### PR TITLE
Diary Implementation

### PR DESCRIPTION
**Pull request by Hroy02**

### CODE REVIEW

File: src/main/java/bgu/spl/mics/application/passiveObjects/Diary.java [hunk 1]
```diff
@@ -11,24 +24,36 @@
...
 	 */
 	public static Diary getInstance() {
-		//TODO: Implement this
-		return null;
+		return _instance;
+	}
+
+	private Diary() {
+		_diary = new ArrayList<>();
+		_lock = new ReentrantReadWriteLock();
```
Comment (IdanTo1) [inline src/main/java/bgu/spl/mics/application/passiveObjects/Diary.java]: please explain what's that and why do you use it
a. to me :)
b. as a comment in the code

File: src/main/java/bgu/spl/mics/application/passiveObjects/Diary.java [hunk 2]
```diff
@@ -38,23 +63,33 @@ public void addReport(Report reportToAdd){
 	 * List of all the reports in the diary.
 	 * This method is called by the main method in order to generate the output.
 	 */
-	public void printToFile(String filename){
-		//TODO: Implement this
+	public void printToFile(String filename) throws IOException {
+		_lock.readLock().lock();
+		Gson gson = new Gson();
+		String reportsToPrint = gson.toJson(_diary);
+		PrintWriter writer = null;
```
Comment (IdanTo1) [inline src/main/java/bgu/spl/mics/application/passiveObjects/Diary.java]: ?

File: src/main/java/bgu/spl/mics/application/passiveObjects/Diary.java [hunk 3]
```diff
@@ -38,23 +63,33 @@ public void addReport(Report reportToAdd){
...
-		return 0;
+		return _total.get();
 	}
 
 	/**
 	 * Increments the total number of received missions by 1
 	 */
 	public void incrementTotal(){
-
+		int localTotal;
```
Comment (IdanTo1) [inline src/main/java/bgu/spl/mics/application/passiveObjects/Diary.java]: call it oldTotal, to  stick with convention

File: src/main/java/bgu/spl/mics/application/passiveObjects/Diary.java [hunk 4]
```diff
@@ -38,23 +63,33 @@ public void addReport(Report reportToAdd){
...
+		do
+		{
+			localTotal = _total.get();
+		}
+		while (!_total.compareAndSet(localTotal, localTotal+1));
```
Comment (IdanTo1) [inline src/main/java/bgu/spl/mics/application/passiveObjects/Diary.java]: is this +1 safe? maybe save a  newTotal variable in the do loop

**Review by IdanTo1 — CHANGES_REQUESTED** [review]
looking very good
just explain better what you've done
and minor fixes

File: src/main/java/bgu/spl/mics/application/passiveObjects/Diary.java (hunk 1 above)
Comment (Hroy02) [reply src/main/java/bgu/spl/mics/application/passiveObjects/Diary.java]: I am adding a comment. It is also explained [here](https://www.cs.bgu.ac.il/~spl201/index.php?page=Thread-safty2#readerswriters_example)

File: src/main/java/bgu/spl/mics/application/passiveObjects/Diary.java (hunk 2 above)
Comment (Hroy02) [reply src/main/java/bgu/spl/mics/application/passiveObjects/Diary.java]: Considering I straight up stole it from your Inventory implementation you should probably upload an update to it :)

File: src/main/java/bgu/spl/mics/application/passiveObjects/Diary.java (hunk 3 above)
Comment (Hroy02) [reply src/main/java/bgu/spl/mics/application/passiveObjects/Diary.java]: Done

File: src/main/java/bgu/spl/mics/application/passiveObjects/Diary.java (hunk 4 above)
Comment (Hroy02) [reply src/main/java/bgu/spl/mics/application/passiveObjects/Diary.java]: It is run before calling compareAndSet, so if some thread comes in between them the cas will still fail. This won't help against a thread coming in between but it isn't different by any means than putting it inside the do-while

File: src/main/java/bgu/spl/mics/application/passiveObjects/Diary.java [hunk 5]
```diff
@@ -38,22 +39,24 @@ public static Diary getInstance() {
 
 	private Diary() {
 		_diary = new ArrayList<>();
-		_lock = new ReentrantReadWriteLock();
-		_total = new AtomicInteger(0);
+		// Using Read-Write Lock to safeguard the diary from multiple threads accessing it at the same time
```
Comment (IdanTo1) [inline src/main/java/bgu/spl/mics/application/passiveObjects/Diary.java]: this comment is redundant

File: src/main/java/bgu/spl/mics/application/passiveObjects/Diary.java [hunk 6]
```diff
@@ -63,33 +66,25 @@ public void addReport(Report reportToAdd){
...
 	 */
-	public void incrementTotal(){
-		int localTotal;
-		do
-		{
-			localTotal = _total.get();
-		}
-		while (!_total.compareAndSet(localTotal, localTotal+1));
+	public synchronized void incrementTotal(){
+		_total++;
```
Comment (IdanTo1) [inline src/main/java/bgu/spl/mics/application/passiveObjects/Diary.java]: I think this one is better with the AtomicInteger. no need to synchornize here. synchornize only when adding, printing, or getting the diary, and lock on this object..

**Review by IdanTo1 — CHANGES_REQUESTED** [review]
pay attention to the comments, no extremity needed, meet me in the middle (:

**Review by IdanTo1 — DISMISSED** [review]

**Review by IdanTo1 — APPROVED** [review]

**Pull request merged**